src/wp-includes/speculative-loading.php

```diff
@@ -0,0 +1,245 @@
+		$prefixer->prefix_path_pattern( '/wp-admin/*', 'site' ),
+		$prefixer->prefix_path_pattern( '/*\\?*(^|&)_wpnonce=*', 'home' ),
+		$prefixer->prefix_path_pattern( '/*', 'uploads' ),
+		$prefixer->prefix_path_pattern( '/*', 'content' ),
+		$prefixer->prefix_path_pattern( '/*', 'plugins' ),
+		$prefixer->prefix_path_pattern( '/*', 'template' ),
+		$prefixer->prefix_path_pattern( '/*', 'stylesheet' ),
+	);
+
+	/**
+	 * Filters the paths for which speculative loading should be disabled.
+	 *
+	 * All paths should start in a forward slash, relative to the root document. The `*` can be used as a wildcard.
+	 * If the WordPress site is in a subdirectory, the exclude paths will automatically be prefixed as necessary.
+	 *
+	 * Note that WordPress always excludes certain path patterns such as `/wp-login.php` and `/wp-admin/*`, and those
+	 * cannot be modified using the filter.
+	 *
+	 * @since 6.8.0
+	 *
+	 * @param string[] $href_exclude_paths Additional path patterns to disable speculative loading for.
+	 * @param string   $mode               Mode used to apply speculative loading. Either 'prefetch' or 'prerender'.
+	 */
+	$href_exclude_paths = (array) apply_filters( 'wp_speculation_rules_href_exclude_paths', array(), $mode );
+
+	// Ensure that:
+	// 1. There are no duplicates.
+	// 2. The base paths cannot be removed.
+	// 3. The array has sequential keys (i.e. array_is_list()).
+	$href_exclude_paths = array_values(
+		array_unique(
+			array_merge(
+				$base_href_exclude_paths,
+				array_map(
+					static function ( string $href_exclude_path ) use ( $prefixer ): string {
+						return $prefixer->prefix_path_pattern( $href_exclude_path );
+					},
+					$href_exclude_paths
+				)
+			)
+		)
+	);
+
+	$rules = array(
+		array(
+			'source'    => 'document',
+			'where'     => array(
+				'and' => array(
+					// Include any URLs within the same site.
+					array(
+						'href_matches' => $prefixer->prefix_path_pattern( '/*' ),
+					),
+					// Except for excluded paths.
+					array(
+						'not' => array(
+							'href_matches' => $href_exclude_paths,
+						),
+					),
+					// Also exclude rel=nofollow links, as certain plugins use that on their links that perform an action.
+					array(
+						'not' => array(
+							'selector_matches' => 'a[rel~="nofollow"]',
+						),
+					),
+					// Last but not least, exclude links that are explicitly marked to opt out.
+					array(
+						'not' => array(
+							'selector_matches' => ".no-{$mode}",
+						),
+					),
+				),
+			),
+			'eagerness' => $eagerness,
+		),
+	);
+
+	return array( $mode => $rules );
+}
+
+/**
+ * Prints the speculation rules.
+ *
+ * For browsers that do not support speculation rules yet, the `script[type="speculationrules"]` tag will be ignored.
+ *
+ * @since 6.8.0
+ * @access private
+ */
+function wp_print_speculation_rules(): void {
+	$configuration = wp_get_speculation_rules_configuration();
+	if ( false === $configuration ) {
```

Comment:
```suggestion
	if ( ! is_array( $configuration ) ) {
```

Reply:
See above, I think an explicit check for `false` (or `null`) makes more sense. It should not be possible to return any other values.

Reply:
Things are always possible in WordPress 😄

Reply:
Right, as `wp_get_speculation_rules_configuration()` would be typed to return either an `array` or `false`/`null`. In the latter case:

```suggestion
	if ( null === $configuration ) {
```

Reply:
Makes sense. I'll update this on Monday, as it will require test changes too.

Reply:
Updated in https://github.com/WordPress/wordpress-develop/pull/7860/commits/cc66024af13833c2a4ae1704e6de8a49ca766feb